compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,608 @@
+    // this could end up duplicating the transposes. We should only propagate
+    // through reshape when it is free to do so.
+    if (!transposeOp || !transposeOp->hasOneUse()) {
+      return failure();
+    }
+
+    ArrayRef<int64_t> perm = transposeOp.getPermutation();
+
+    auto invPerm = invertPermutationVector(perm);
+    SmallVector<ReassociationIndices> reassociations =
+        expandOp.getReassociationIndices();
+
+    // Because we are doing expand_shape(transpose), all expanded groups are
+    // transposed together. As a result, to get the permutation of the new
+    // transpose, we can just flatten the transposed reassociation indices.
+    // For example,
+    //
+    // permutation = [0, 2, 1]
+    // reassociation_map = [[0, 1, 2], [3], [4, 5]]
+    //
+    // Becomes
+    //
+    // reassociation_map = [[0, 1, 2], [4, 5], [3]]
+    // permutation = [0, 1, 2, 3, 4, 5]
+    applyPermutationToVector(reassociations, invPerm);
+
+    SmallVector<int64_t> newInvPerm;
+    SmallVector<ReassociationIndices> newReassociations;
+    int64_t expandedDim = 0;
+    for (auto reassoc : reassociations) {
+      ReassociationIndices newReassoc;
+      for (auto dim : reassoc) {
+        newInvPerm.push_back(dim);
+        newReassoc.push_back(expandedDim++);
+      }
+      newReassociations.push_back(newReassoc);
+    }
+
+    auto newPerm = invertPermutationVector(newInvPerm);
+
+    RankedTensorType expandedType = getPermutedTensorType(
+        cast<RankedTensorType>(expandOp.getType()), newInvPerm);
+    Value transposedReshape = rewriter.create<tensor::ExpandShapeOp>(
+        expandOp.getLoc(), expandedType, transposeOp.getInput(),
+        newReassociations);
+    Value originalReshape =
+        createTranspose(rewriter, transposedReshape, newPerm);
+    rewriter.replaceOp(expandOp, originalReshape);
+    return success();
+  }
+};
+
+// Fuses a transpose with the input of a linalg.generic op. For example,
+//
+// %0 = linalg.transpose ... permutation = [0, 2, 1] :
+//                           tensor<2x5x7> -> tensor<2x7x5>
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>)
+//
+// Becomes
+//
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d2, d0)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x5x7>)
+//
+// This is considered just one way to model transpose propagation to generics,
+// another option would be to interpret the transpose on the iterators of the
+// generic, thus producing a transpose on the output and any other inputs to
+// the generic. This has the potential introduce more transposes/data movement
+// and isn't the way this pass is modeled. Global data layout transformations
+// like that are better suited for pack/unpack propagation rooted on specific
+// operations.
+class FuseTransposeWithGenericConsumer
```

Comment:
I thought this is modeled by FusionOnTensors at flow level?

Reply:
It is, but I'm thinking of three main reasons for this.
1. This eases the analysis in FusionOnTensors as we want to do fusions like this anyway.
2. In a propagation pass like this, the goal is to propagate the transpose to a "terminal" location, and if we fail then we propagate in the other direction. The current implementation doesn't include anything in "the other direction" so this does look a little strange right now though.
3. This pass is definitely not an optimal propagation, but the idea is that data-tiling (or rather a generalization of set-encoding) should be responsible for resolving preferred layouts. By fusing such transposes in advance, in the fullness of time layout selection should be able to make better use of the extra transpose information.